tests/unit/fem/test_face_permutation.cpp

```diff
@@ -0,0 +1,231 @@
+   }
+
+   std::rotate(face_1.begin(), face_1.begin() + rot, face_1.end());
+   std::rotate(face_2.begin(), face_2.begin() + rot, face_2.end());
+
+   for (int i=0; i<4; ++i)
+   {
+      v[i] = face_1[i];
+      v[i+4] = face_2[i];
+   }
+}
+
+Mesh *mesh_3d_orientation(int face_perm_1, int face_perm_2)
+{
+   static const int dim = 3;
+   static const int nv = 12;
+   static const int nel = 2;
+   Mesh *mesh = new Mesh(dim, nv, nel);
+   double x[dim];
+   x[0] = 0.0;   x[1] = 0.0;   x[2] = 0.0;
+   mesh->AddVertex(x);
+   x[0] = 1.0;   x[1] = 0.0;   x[2] = 0.0;
+   mesh->AddVertex(x);
+   x[0] = 2.0;   x[1] = 0.0;   x[2] = 0.0;
+   mesh->AddVertex(x);
+   x[0] = 0.0;   x[1] = 1.0;   x[2] = 0.0;
+   mesh->AddVertex(x);
+   x[0] = 1.0;   x[1] = 1.0;   x[2] = 0.0;
+   mesh->AddVertex(x);
+   x[0] = 2.0;   x[1] = 1.0;   x[2] = 0.0;
+   mesh->AddVertex(x);
+   x[0] = 0.0;   x[1] = 0.0;   x[2] = 1.0;
+   mesh->AddVertex(x);
+   x[0] = 1.0;   x[1] = 0.0;   x[2] = 1.0;
+   mesh->AddVertex(x);
+   x[0] = 2.0;   x[1] = 0.0;   x[2] = 1.0;
+   mesh->AddVertex(x);
+   x[0] = 0.0;   x[1] = 1.0;   x[2] = 1.0;
+   mesh->AddVertex(x);
+   x[0] = 1.0;   x[1] = 1.0;   x[2] = 1.0;
+   mesh->AddVertex(x);
+   x[0] = 3.0;   x[1] = 1.0;   x[2] = 1.0;
+   mesh->AddVertex(x);
+
+   int el[8];
+
+   el[0] = 0;
+   el[1] = 1;
+   el[2] = 4;
+   el[3] = 3;
+   el[4] = 6;
+   el[5] = 7;
+   el[6] = 10;
+   el[7] = 9;
+   rotate_3d_vertices(el, face_perm_1/4, face_perm_1%4);
+   mesh->AddHex(el);
+
+   el[0] = 1;
+   el[1] = 2;
+   el[2] = 5;
+   el[3] = 4;
+   el[4] = 7;
+   el[5] = 8;
+   el[6] = 11;
+   el[7] = 10;
+   rotate_3d_vertices(el, face_perm_2/4, face_perm_2%4);
+   mesh->AddHex(el);
+
+   mesh->FinalizeHexMesh(true);
+   mesh->GenerateBoundaryElements();
+   mesh->Finalize();
+   return mesh;
+}
+
+double x_fn(const Vector &xvec) { return xvec[0]; }
+double y_fn(const Vector &xvec) { return xvec[1]; }
+double z_fn(const Vector &xvec) { return xvec[2]; }
+
+void TestFaceRestriction(Mesh &mesh, int order)
+{
+   int dim = mesh.Dimension();
+   DG_FECollection fec(order, dim, BasisType::GaussLobatto);
+   FiniteElementSpace fes(&mesh, &fec);
+   GridFunction gf(&fes);
+
+   L2FaceRestriction restr(fes, ElementDofOrdering::LEXICOGRAPHIC,
+                           FaceType::Interior, L2FaceValues::DoubleValued);
+
+   const int ndof_face = pow(order+1, dim-1);
+
+   Vector face_values(ndof_face*2);
+
+   for (int d=0; d<dim; ++d)
+   {
+      double (*fn)(const Vector &);
+      if (d == 0)
+      {
+         fn = x_fn;
+      }
+      else if (d == 1)
+      {
+         fn = y_fn;
+      }
+      else if (d == 2)
+      {
+         fn = z_fn;
+      }
+      else
+      {
+         MFEM_ABORT("Bad dimension");
+         return;
+      }
+      FunctionCoefficient coeff(fn);
+      gf.ProjectCoefficient(coeff);
+      restr.Mult(gf, face_values);
+
+      for (int i=0; i<ndof_face; ++i)
+      {
+         REQUIRE(face_values[i] == face_values(i + ndof_face));
```

Comment:
Brackets [ ] ( )

Reply:
I'm not sure I understand why it compiled with `[]` since `vector` doesn't seem to have this operator. I made the change.

Reply:
I guess it's doing an implicit cast to `double *` first but that way you don't get bounds checking in debug mode.